### ISSUE TITLE
Add `immer` to `createReducerFromClass`

### ISSUE THREAD

**Issue by HarelM** [angular-redux2/store#10]
### This is a...

* [x] feature request
* [ ] bug report
* [ ] usage question

### What toolchain are you using for transpilation/bundling?

* [x] @angular/cli
* [ ] Custom @ngTools/webpack
* [ ] Raw `ngc`
* [ ] SystemJS
* [ ] Rollup
* [ ] Other

### Environment

NodeJS Version: 16
Typescript Version: 4.7
Angular Version:
@angular-redux/store version:
@angular/cli version: (if applicable)
OS:

### Additional Notes:
Since creating a reducer from a class already generates code it can be super beneficial to introduce immer there and allow the developer to reduce the boilerplate code even further.

This is used in redux-toolkit verify similar to how it's done here...

Let me know if this is something you would like me to push forward by sending a PR

**Comment by Garefild** [comment]
Hi, I'm working on it right now, 
if you have any insight I'd love to hear it.
Hope to release a version in the next few hours. 
with a few more improvements.

**Comment by HarelM** [comment]
Haven't got to POC it, just read about it, and it sounded super cool and efficient...

**Comment by Garefild** [comment]
Yepp, If there are more things to add, I would appreciate suggestions.

**Comment by HarelM** [comment]
Well, currently, the only thing that still annoys me (assuming this will be introduced) when I'm adding things to the redux state is the action and payload class/type generation.
But I don't know how to simplify this boilerplate code...
I mean every time I would like to create a new state update I need to create
1. Payload type
2. Action class
3. A method in the reducer

While there's no argue 3 is needed I would be glad to reduce the code related to 1 & 2, but I have no clue how...

**Comment by Garefild** [comment]
Omm ok, I will see what I can do to minimize this.

**Comment by HarelM** [comment]
Here is an example of a fairly large state (but not that large):
https://github.com/IsraelHikingMap/Site/blob/06e21d239f4567f3beb06ff6c848cbf7fac4d7a9/IsraelHiking.Web/src/application/reducers/routes.reducer.ts
You can see how much boiler plate code there is there... :-(

**Comment by Garefild** [comment]
Oh wow. 
It's will take me more than a few hours to release a version.
I think I have a solution.

**Comment by Garefild** [comment]
But I need to see how I leave support for 4.0.4 so you don't have to change everything you've already written.

**Comment by HarelM** [comment]
I don't mind refactoring all my reducers to reduce the amount of code.
In any case, let me know when I can tests things... Thanks a lot for all the hard work!

**Comment by Garefild** [comment]
Ok, I will send msg here when I finish the version u able to test.

**Comment by Garefild** [comment]
This is how it looks in the new approach.

```typescript
import undoable, { UndoableOptions, includeAction, groupByActionTypes } from "redux-undo";
import { Action, AnyAction } from "@angular-redux2/store";

import { initialState, BaseAction } from "./initial-state";
import type { RouteData, MarkerData, RouteSegmentData, RouteStateName, LatLngAltTime } from "../models/models";

export class RoutesReducer extends AbstractReducer {
    private doForRoute(lastState: RouteData[], routeId: string, updateAction: (routeToUpdate: RouteData) => RouteData): RouteData[] {
        let route = lastState.find(r => r.id === routeId);
        let routes = [ ...lastState ];
        let updatedRoute = updateAction(route);
        routes.splice(routes.indexOf(route), 1, updatedRoute);
        return routes;
    }

    @Action
    public addRoute(lastState: RouteData[], action: AnyAction): RouteData[] {
        return [ ...lastState, action.payload.routeData ];
    }

    @Action
    public deleteRoute(lastState: RouteData[], action: AnyAction): RouteData[] {
        let routes = [ ...lastState ];
        let routeToRemove = routes.find(r => r.id === action.payload.routeId);
        routes.splice(routes.indexOf(routeToRemove), 1);
        return routes;
    }

    @Action
    public changeProperties(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState, action.payload.routeId,
            (route) =>
                ({
                    ...route,
                    name: action.payload.routeData.name,
                    opacity: action.payload.routeData.opacity || route.opacity,
                    weight: action.payload.routeData.weight || route.weight,
                    color: action.payload.routeData.color || route.color,
                    description: action.payload.routeData.description || route.description,
                } as RouteData));
    }

    @Action
    public addPoi(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState, action.payload.routeId,
            (route) =>
                ({
                    ...route,
                    markers: [ ...route.markers, action.payload.markerData ]
                } as RouteData));
    }

    @Action
    public updatePoi(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState,
            action.payload.routeId,
            (route) => {
                let markers = [ ...route.markers ];
                markers.splice(action.payload.index, 1, action.payload.markerData);
                return {
                    ...route,
                    markers
                } as RouteData;
            });
    }

    @Action
    public deletePoi(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState,
            action.payload.routeId,
            (route) => {
                let markers = [ ...route.markers ];
                markers.splice(action.payload.index, 1);
                return {
                    ...route,
                    markers
                } as RouteData;
            });
    }

    @Action
    public addSegment(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState,
            action.payload.routeId,
            (route) =>
                ({
                    ...route,
                    segments: [ ...route.segments, action.payload.segmentData ]
                } as RouteData));
    }

    @Action
    public updateSegment(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState,
            action.payload.routeId,
            (route) => {
                let segments = [ ...route.segments ];
                if (action.payload.segmentsData.length === action.payload.indices.length) {
                    for (let segmentIndex = 0; segmentIndex < action.payload.indices.length; segmentIndex++) {
                        segments.splice(action.payload.indices[segmentIndex], 1, action.payload.segmentsData[segmentIndex]);
                    }
                } else if (action.payload.segmentsData.length === 2 && action.payload.indices.length === 1) {
                    segments.splice(action.payload.indices[0], 1, ...action.payload.segmentsData);
                } else if (action.payload.segmentsData.length === 1 && action.payload.indices.length === 2) {
                    segments.splice(action.payload.indices[1], 1, action.payload.segmentsData[0]);
                    segments.splice(action.payload.indices[0], 1);
                }
                return {
                    ...route,
                    segments
                } as RouteData;
            });
    }

    @Action
    public replaceSegments(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState,
            action.payload.routeId,
            (route) => ({
                    ...route,
                    segments: action.payload.segmentsData
                } as RouteData
            ));
    }

    @Action
    public deleteSegment(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState,
            action.payload.routeId,
            (route) => {
                let segments = [ ...route.segments ];
                segments.splice(action.payload.index, 1);
                return {
                    ...route,
                    segments
                } as RouteData;
            });
    }

    @Action
    public changeEditState(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState,
            action.payload.routeId,
            (route) =>
                ({
                    ...route,
                    state: action.payload.state
                } as RouteData));
    }

    @Action
    public changeVisibility(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState,
            action.payload.routeId,
            (route) =>
                ({
                    ...route,
                    state: action.payload.isVisible ? "ReadOnly" : "Hidden"
                } as RouteData));
    }

    @Action
    public reverseRoute(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState,
            action.payload.routeId,
            () => action.payload.routeData);
    }

    @Action
    public splitRoute(lastState: RouteData[], action: AnyAction): RouteData[] {
        let route = lastState.find(r => r.id === action.payload.routeId);
        let routes = [ ...lastState ];
        routes.splice(routes.indexOf(route), 1, action.payload.routeData, action.payload.splitRouteData);
        return routes;
    }

    @Action
    public mergeRoutes(lastState: RouteData[], action: AnyAction): RouteData[] {
        let route = lastState.find(r => r.id === action.payload.routeId);
        let secondaryRoute = lastState.find(r => r.id === action.payload.secondaryRouteId);
        let routes = [ ...lastState ];
        routes.splice(routes.indexOf(route), 1, action.payload.mergedRouteData);
        routes.splice(routes.indexOf(secondaryRoute), 1);
        return routes;
    }

    @Action
    public addRecordingPoint(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState,
            action.payload.routeId,
            (route) => {
                let segments = [ ...route.segments ];
                let lastSegment = { ...segments[segments.length - 1] };
                lastSegment.latlngs = [ ...lastSegment.latlngs, ...action.payload.latlngs ];
                lastSegment.routePoint = action.payload.latlngs[action.payload.latlngs.length - 1];
                segments.splice(segments.length - 1, 1, lastSegment);
                return {
                    ...route,
                    segments
                };
            });
    }

    @Action
    public clearPois(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState,
            action.payload.routeId,
            (route) => ({
                ...route,
                markers: []
            }));
    }

    @Action
    public clearPoisAndRoute(lastState: RouteData[], action: AnyAction): RouteData[] {
        return this.doForRoute(lastState,
            action.payload.routeId,
            (route) => ({
                ...route,
                markers: [],
                segments: []
            }));
    }

    @Action
    public deleteAllRoutes(lastState: RouteData[], action: AnyAction): RouteData[] {
        return [];
    }

    @Action
    public toggleAllRoutes(lastState: RouteData[], action: AnyAction): RouteData[] {
        let routes = [ ...lastState ];
        let isAllRoutesHidden = routes.find(r => r.state !== "Hidden") == null;
        for (let route of routes) {
            route.state = isAllRoutesHidden ? "ReadOnly" : "Hidden";
        }
        return routes;
    }

    @Action
    public replaceRoutes(lastState: RouteData[], action: AnyAction): RouteData[] {
        return action.payload.routesData;
    }
}

export const routesReducer = RoutesReducer.createReducer(initialState.routes.present);
```

```typescript
import { Component, ViewEncapsulation, Inject } from "@angular/core";
import { MAT_DIALOG_DATA } from "@angular/material/dialog";
import { NgRedux } from "@angular-redux2/store";

import { RouteBaseDialogComponent } from "./route-base-dialog.component";
import { ResourcesService } from "../../../services/resources.service";
import { FileService } from "../../../services/file.service";
import { FitBoundsService } from "../../../services/fit-bounds.service";
import { ToastService } from "../../../services/toast.service";
import { RoutesFactory } from "../../../services/routes.factory";
import { SelectedRouteService } from "../../../services/selected-route.service";
import { SpatialService } from "../../../services/spatial.service";

import { SetSelectedRouteAction } from "application/reducers/route-editing-state.reducer";
import type { DataContainer, RouteData, ApplicationState, LatLngAlt } from "../../../models/models";

// NEW
import { RoutesReducer } from "../../../reducers/routes.reducer";

@Component({
    selector: "route-edit-dialog",
    templateUrl: "./route-properties-dialog.component.html",
    styleUrls: [ "./route-properties-dialog.component.scss" ],
    encapsulation: ViewEncapsulation.None,
})
export class RouteEditDialogComponent extends RouteBaseDialogComponent {
    private originalName: string;

    constructor(resources: ResourcesService,
                selectedRouteService: SelectedRouteService,
                routesFactory: RoutesFactory,
                toastService: ToastService,
                ngRedux: NgRedux<ApplicationState>,
                private readonly fileService: FileService,
                private readonly fitBoundsService: FitBoundsService,
                @Inject(MAT_DIALOG_DATA) data: RouteData
    ) {
        super(resources, selectedRouteService, routesFactory, toastService, ngRedux);

        this.isNew = false;
        this.title = this.resources.routeProperties;
        this.routeData = data;
        this.originalName = data.name;
    }

    public deleteRoute() {
        this.ngRedux.dispatch(
            RoutesReducer.actions.deleteRoute({
                routeId: this.routeData.id
            })
        );
    }
}
```

**Comment by Garefild** [comment]
Waiting to receive feedback from you, is this what you wanted? (to know if I'm on the right track)
I didn't delve into your code, but why don't you use a sub-store?
Or should I develop something dedicated to routers?
Why exactly do you use redux in routers?

Shabbat Shalom.

**Comment by HarelM** [comment]
This is an interesting approach.
Routes in this case are ways and points, not urls.
Not sure what you mean about sub store though.
Is the `actions` field auto generated?
`AnyAction` can probably be improved, but if I can inline type it in the methods it will be great.

If all this will work and generate the relevant types and allow auto complete it will be great.

Also note that there's no reason to do all the "..." things when using immer, so it will be even less code.

**Comment by Garefild** [comment]
"This is an interesting approach." - txn :) 
I spent a lot of time thinking about an approach that would not compromise performance.

"Routes in this case are ways and points, not URLs."
Can you explain a little about it?

"Not sure what you mean about sub store though."
You can break a large object in a store into several small objects that work together.
Then it's faster to create a new object than to duplicate a huge object.
I will write you an example when I finish the improvements.

Is the actions field auto generated?
yepp

"AnyAction can probably be improved, but if I can inline type it in the methods it will be great.
If all this will work and generate the relevant types and allow auto complete it will be great."
I'm working on it, it's a bit more complex.


"Also note that there's no reason to do all the "..." things when using immer, so it will be even less code."
I've added it to the code, I'll explain later how to duplicate an object in an efficient way that doesn't hurt performance.

**Comment by HarelM** [comment]
>"Routes in this case are ways and points, not URLs."
Can you explain a little about it?

![image](https://user-images.githubusercontent.com/3269297/197360054-229b468a-d3aa-4c21-ba36-76254ad5d697.png)
The above image shows the data structure I'm using on a map - route as is a route you can follow, physically - a way, path, track, trace, you name it, it's the same thing.

I'm not sure I can actually break this part of the object to a smaller one - this entire "routes" list is an undoable list that gets updates when a user edits its routes, every edit a new list is created in order to save undo and redo history - this entire list is a part of a state so I'm not sure I can really use a sub-store here. The state in general is not a very deep object so I'm not sure I need a sub store, but I might be wrong... :-)

You can, in theory play with it here:
https://israelhiking.osm.org.il/
But it's a bit of a hassle to learn, so I'm not expecting you'll actually go there :-)

Overall I like the reduction in boilerplate code if everything goes as you plan.

**Comment by Garefild** [comment]
In the case of "Immer" I made a proxy object so u can use it like this:

```typescript
    @Action
    isLogin(state: Users, action: any) {
        state['alice'].name = Math.random();

        return state;
    }
```

u apply the change to the state and the proxy makes a new object.

**Comment by HarelM** [comment]
Great!

**Comment by Garefild** [comment]
> Great!

I didn't forget, I need to redesign large parts of the system.
test performance and backward compatibility.
and auto-completion support for action (type/interface).

**Comment by Garefild** [comment]
Hi, 5.0.0-beta.1 has publish :)
https://www.npmjs.com/package/@angular-redux2/store/v/5.0.0-beta.1

**Comment by HarelM** [comment]
Is `immer` being used? I don't see it in the code...

**Comment by HarelM** [comment]
I've started migration my code to use the new actions, and it seems that the object that is sent to the reducer class is not an `Action` object with a payload but rather the payload object.
I'm not sure if this was intended or not.
According to what you wrote here: https://github.com/angular-redux2/store/issues/9#issuecomment-1335783337 this wasn't intended.
Waiting on your response to continue with the migration process of my code.
Thanks for all the hard work!

**Comment by HarelM** [comment]
Probably this is due to this line:
https://github.com/angular-redux2/store/blob/585d80c92a96b7298a9b6c2ff8e60a232c378d1e/projects/angular-redux2/store/src/abstract/reducer.abstract.ts#L77
Which takes the payload from the action and sends it to the function, if I need to guess...

**Comment by HarelM** [comment]
I've created a branch in my code so that I can test the latest version.
I did all the changes in my code, but I'll test it only after the above is fixed or decided at least.
Code can be found here:
https://github.com/IsraelHikingMap/Site/tree/update-redux-store

I think the only thing that still bugs me is the fact that I need the `actions` and the method names to be the same, and refactoring it without being careful might cause the reducer to break, which is less than ideal.
I'll see if I can look at the code and maybe think of a smart way to allow auto-complete (as I think the code is being generated appropriately, right?)

**Comment by Garefild** [comment]
> Is `immer` being used? I don't see it in the code...

No, you just useit like immer.

```typescript

function reducer(state: any, action: any): void{
    state.name = 'xxx';
}

```

**Comment by HarelM** [comment]
I'm probably missing something basic - if this library doesn't use immer how will the state object kept immutable?

**Comment by HarelM** [comment]
For the type stuff I tried the following but it didn't work, not sure why though:
```ts
type ArgumentTypes<F> = F extends (state: any, action: AnyAction<infer A>) => any ? A : never;
export type ReduceActions<TType> = {[K in keyof TType]: ActionPayload<ArgumentTypes<TType[K]>>};

class TestReducer extends AbstractReducer {
    /**
     * Action types
     */

    static override actions: ReduceActions<TestReducer>

    addBug(state: Array<BugPayload>, action: AnyAction<BugPayload>): Array<BugPayload> {
        const payload: BugPayload = action['payload'];
        return [ payload ];
    }
}
```

Hovering on `actions` doesn't give the right type unfortunately:
![image](https://user-images.githubusercontent.com/3269297/206563862-66301f87-f046-4dc6-8d82-72ca24a383db.png)

**Comment by Garefild** [comment]
> I've started migration my code to use the new actions, and it seems that the object that is sent to the reducer class is not an > Action object with a payload but rather the payload object. 

it's action with payload object (the action generate behind the scenes).

> I think the only thing that still bugs me is the fact that I need the actions and the method names to be the same, and > refactoring it without being careful might cause the reducer to break, which is less than ideal.

Yes and No 
when u using the @Action decorator it generates an anonymous function that gets a payload if necessary.
the *static actions* it's pure for auto-complete.

**Comment by HarelM** [comment]
Can you take a look at the following reducer:
https://github.com/IsraelHikingMap/Site/blob/1b3c4d15d30f064fc6f59673fcce9d1c6322b929/IsraelHiking.Web/src/application/reducers/configuration.reducer.ts#L55
When I use it, it says that `language` is undefined.
This is because the method receives the payload object and not the action object.
I can change my code to assume I'm getting the payload object and not the action object, but I'm just not sure if this is what you planned or is this a bug...

**Comment by Garefild** [comment]
That was my mistake, I should have updated you on last-minute changes.
After I changed your web it looked to me very annoying to do AnyAction<SetSelectedPoiPayload> so I changed it to SetSelectedPoiPayload.


![image](https://user-images.githubusercontent.com/2126092/206567216-126d1495-e15e-4dd3-867d-043b9fa79f14.png)

to this

![image](https://user-images.githubusercontent.com/2126092/206567406-18228439-9c5e-45a2-8546-48d6e9e1cc96.png)

**Comment by Garefild** [comment]
I might undo this change.
Tell me what you think.

**Comment by HarelM** [comment]
It saves writing redundant code, it's ok.
I'll change my code tomorrow and do some testing, maybe I'll be able to make the right type for the actions object to reduce the classes even more...

**Comment by Garefild** [comment]
If you want I open the discord group.
https://discord.gg/cSRtr4Wv

**Comment by HarelM** [comment]
I've updated my code, in most cases the state seems to work as expected.
Thanks for all the hard work!
Having said that, the `undoable` reducer I'm using is not working as expected.
I believe this is due to the fact that I'm updating the state as if it was using `immer` but intact I'm updating the previous state, which doesn't work well with `redux-undo`.
The following reducer can be found here:
https://github.com/IsraelHikingMap/Site/blob/ab30f4481e85b668acd2b2c8e362c60a30b212e1/IsraelHiking.Web/src/application/reducers/routes.reducer.ts
I generally believe that `immer` is needed here, I would be happy to know that I'm wrong though :-)

**Comment by HarelM** [comment]
Ahh! Finally got the types working!
The following basically reduces the need to type all the actions:
```ts
type Args<T> = T extends (...args: infer X) => any ? X : never;
type ReducerActions<T> = {[K in keyof T]: ActionPayload<Args<T[K]>[1]>}
...
export class InMemoryReducer extends AbstractReducer {
    static actions: ReducerActions<InMemoryReducer>;
...
}

```

Feel free to add it here so it can be easily used to reduce the code even more and avoid problems when refactoring.
This can't be defined in the abstract class due to static usage and problem with generic. But I think it's good enough this way...

**Comment by Garefild** [comment]
> I generally believe that immer is needed here, I would be happy to know that I'm wrong though :-)

`Immer` - Adds complications and slows down the system.
And you don't need all of its utilities, so I used a native js proxy object to detect changes and create a new object behind the scenes.

I would appreciate it if you could direct me to what is not working so that I can check, maybe my design was not correct.

**Comment by Garefild** [comment]
> ``` ts
> type Args<T> = T extends (...args: infer X) => any ? X : never;
> type ReducerActions<T> = {[K in keyof T]: ActionPayload<Args<T[K]>[1]>}
> ...
> export class InMemoryReducer extends AbstractReducer {
>     static actions: ReducerActions<InMemoryReducer>;
> ...
> }
> ```

Umm what if u have a none reducer function? 
Good idea let's see if I can remove entirely the `static actions`.

**Comment by HarelM** [comment]
I'll try and reproduce a stackblitz example later tonight to show the problem, assuming I can reproduce this.

If you have a non reducer method you can always type it "manually" I guess...

**Comment by Garefild** [comment]
[feature/5.0.0-beta.2](https://github.com/angular-redux2/store/tree/feature/5.0.0-beta.2)
 `ReducerActions` add to the project and released in the next beta.

**Comment by Garefild** [comment]
> 'll try and reproduce a stackblitz example later tonight to show the problem, assuming I can reproduce this.

When you working on this and if I'm offline send me a message on discord the notification will alert me.
the link in the readme file and above.

**Comment by HarelM** [comment]
I think I managed to reproduce the problem here:
https://stackblitz.com/edit/angular-ivy-yo4xs1?file=src/app/app.component.ts
check out the reducer function.
I'm on and off the computer due to kids, maybe later tonight I'll be available...

**Comment by HarelM** [comment]
> Immer - Adds complications and slows down the system.
And you don't need all of its utilities, so I used a native js proxy object to detect changes and create a new object behind the scenes.
I would appreciate it if you could direct me to what is not working so that I can check, maybe my design was not correct.

While I admire the courage to try and write a small piece of code to detect object changes, I would recommend using something that is more battle tested. 
`immer` has 9M downloads a week, it says something I guess, but if it's too complicated, you might be able to use `on-change` https://www.npmjs.com/package/on-change, it has 25K downloads a week which is nice.

The main problem of writing something here instead of using something that someone else wrote is that it's not the main goal of this library, and even worse, it's extremely hard to see that this causes an issue as redux might just work if you don't keep the state immutable, so it might create hard to fix bugs.
If you want to create a library (or a class in this library) that detects object changes you'll need to make sure you have the right code for it, the right tests and everything, and for me it's a bit out of scope. IDK...

If `immer` or anything else might create a performace issue you can maybe have a flag to enable disable it, this way who ever uses this library can decide if they keep writing their code with all the `...` and making sure it has the right performance or allow the library to do what it needs to reduce code complexity and avoid making foolish errors in the reducers, I know I'm in the last part - I just know I'll make mistakes when updating the state, so I prefer something to do it for me, and I don't mind paying the performance penalty....

The last argument is that redux toolkit is using it under the hood:
https://redux-toolkit.js.org/usage/immer-reducers
And it seems that it has good performance based on this article. 
https://blog.bitsrc.io/using-immer-with-react-a-simple-solutions-for-immutable-states-a6ebb8b0bfa

Bottom line, it's your call, just sharing a few more insights. Any thing you decide is great.
Thanks again for all the hard work!

**Comment by HarelM** [comment]
@Garefild I just saw that you publish version 5.0. Congrats!
The discord channel invite has expired, so I'm not sure if you posted some comments there, but I would be interested to know if the problem with the undoable was solved as part of the 5.0 release?
Thanks for all the hard work!

**Comment by Garefild** [comment]
@HarelM Yo, 
The link to chat is an online badge, you can also use the link https://discord.com/invite/7BnsAqst6W.
`redux-undo` is unmaintained so I create a middleware chain of reducer API as a base for the plugin system.
[middleware.md](https://github.com/angular-redux2/store/blob/main/markdown/middleware.md)

and create basic plugins:
* undo - I will release this weekend version 1.0.0
* sync - release version 1.0.1

**Comment by HarelM** [comment]
Thanks for the update. Looking forward to it.

**Comment by HarelM** [comment]
Ping me when you release that version and I'll see if I can add it to my project. Thanks!

**Comment by Garefild** [comment]
Yo, wait for u in the chat. 
Anyway ng-undo version 1.0.0 work with 5.0.2

**Comment by HarelM** [comment]
The invite doesn't work for me, not sure why.
Can you send a link to the npm package? I can't find it and the readme in the undo is empty...
https://github.com/angular-redux2/undo

![image](https://user-images.githubusercontent.com/3269297/236642870-915b14c8-9bca-492c-a62d-10ef4c8a3967.png)

**Comment by HarelM** [comment]
I've decided to migrate to NGXS and write a very simple undo-redu as part of the relevant state.
Thanks for all the support and best of luck with this library!!

**Comment by Garefild** [comment]
Good luck

**Comment by Garefild** [comment]
> I've decided to migrate to NGXS and write a very simple undo-redu as part of the relevant state. Thanks for all the support and best of luck with this library!!

anyway, I publish the lib if you want to take a look. 
https://github.com/angular-redux2/undo
https://www.npmjs.com/package/@angular-redux2/undo
